pkg/rtc/mediatracksubscriptions.go

```diff
@@ -106,91 +101,31 @@ func (t *MediaTrackSubscriptions) AddSubscriber(sub types.LocalParticipant, wr *
-	}
-
-	var trailer []byte
-	if t.params.MediaTrack.IsEncrypted() {
-		trailer = sub.GetTrailer()
-	}
-
-	downTrack, err := sfu.NewDownTrack(sfu.DowntrackParams{
-		Codecs:                         codecs,
-		Source:                         t.params.MediaTrack.Source(),
-		Receiver:                       wr,
-		BufferFactory:                  sub.GetBufferFactory(),
-		SubID:                          subscriberID,
-		StreamID:                       streamID,
-		MaxTrack:                       maxTrack,
-		PlayoutDelayLimit:              sub.GetPlayoutDelayConfig(),
-		Pacer:                          sub.GetPacer(),
-		Trailer:                        trailer,
-		Logger:                         LoggerWithTrack(sub.GetLogger().WithComponent(sutils.ComponentSub), trackID, t.params.IsRelayed),
-		RTCPWriter:                     sub.WriteSubscriberRTCP,
-		DisableSenderReportPassThrough: sub.GetDisableSenderReportPassThrough(),
-		SupportsCodecChange:            sub.SupportsCodecChange(),
+	subTrack, err := NewSubscribedTrack(SubscribedTrackParams{
+		ReceiverConfig:     t.params.ReceiverConfig,
+		SubscriberConfig:   t.params.SubscriberConfig,
+		Subscriber:         sub,
+		MediaTrack:         t.params.MediaTrack,
+		AdaptiveStream:     sub.GetAdaptiveStream(),
+		Telemetry:          t.params.Telemetry,
+		WrappedReceiver:    wr,
+		IsRelayed:          t.params.IsRelayed,
+		OnDownTrackCreated: t.onDownTrackCreated,
+		OnDownTrackClosed: func(subscriberID livekit.ParticipantID) {
+			t.subscribedTracksMu.Lock()
+			delete(t.subscribedTracks, subscriberID)
+			t.subscribedTracksMu.Unlock()
+		},
+		OnSubscriberMaxQualityChange: t.onSubscriberMaxQualityChange,
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	if t.onDownTrackCreated != nil {
-		t.onDownTrackCreated(downTrack)
-	}
-
-	subTrack := NewSubscribedTrack(SubscribedTrackParams{
-		PublisherID:       t.params.MediaTrack.PublisherID(),
-		PublisherIdentity: t.params.MediaTrack.PublisherIdentity(),
-		PublisherVersion:  t.params.MediaTrack.PublisherVersion(),
-		Subscriber:        sub,
-		MediaTrack:        t.params.MediaTrack,
-		DownTrack:         downTrack,
-		AdaptiveStream:    sub.GetAdaptiveStream(),
-	})
-
-	if !sub.Hidden() {
-		downTrack.OnBindAndConnected(func() {
-			t.params.MediaTrack.OnTrackSubscribed()
-		})
-	}
-
 	// Bind callback can happen from replaceTrack, so set it up early
 	var reusingTransceiver atomic.Bool
 	var dtState sfu.DownTrackState
-	downTrack.OnCodecNegotiated(func(codec webrtc.RTPCodecCapability) {
-		if !wr.DetermineReceiver(codec) {
-			if t.onSubscriberMaxQualityChange != nil {
-				go func() {
-					mimeType := mime.NormalizeMimeType(codec.MimeType)
-					spatial := buffer.GetSpatialLayerForVideoQuality(
-						mimeType,
-						livekit.VideoQuality_HIGH,
-						t.params.MediaTrack.ToProto(),
-					)
-					t.onSubscriberMaxQualityChange(downTrack.SubscriberID(), mimeType, spatial)
-				}()
-			}
-		}
-	})
+	downTrack := subTrack.DownTrack()
```

Comment:
Would like to move all `downTrack` related stuff into `SubscribedTrack`, but it is a bit messy. So, leaving it in here for this round. Will take a look again later.